src/util.rs

```diff
@@ -216,38 +217,202 @@ pub fn decode_leb128_s(data: &[u8]) -> Option<(i64, u8)> {
+    }
+
+    /// Read a `u32` value.
+    #[inline]
+    fn read_u32(&mut self) -> Option<u32> {
+        self.read_pod::<u32>()
+    }
+
+    /// Read a `u64` value.
+    #[inline]
+    fn read_u64(&mut self) -> Option<u64> {
+        self.read_pod::<u64>()
     }
 }
 
-#[inline]
-pub fn decode_udword(data: &[u8]) -> u64 {
-    decode_uword(data) as u64 | ((decode_uword(&data[4..]) as u64) << 32)
+impl<'data> ReadExt<'data> for &'data [u8] {
+    #[inline]
+    fn ensure(&self, len: usize) -> Option<()> {
+        if len > self.len() {
+            return None;
+        }
+        Some(())
+    }
+
+    #[inline]
+    fn read_slice(&mut self, len: usize) -> Option<&'data [u8]> {
+        self.ensure(len)?;
+        let (a, b) = self.split_at(len);
+        *self = b;
+        Some(a)
+    }
+
+    #[inline]
+    fn read_cstr(&mut self) -> Option<&'data CStr> {
+        let idx = self.iter().position(|byte| *byte == b'\0')?;
+        CStr::from_bytes_with_nul(self.read_slice(idx + 1)?).ok()
+    }
+}
+
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+
+    /// Make sure that `[u8]::ensure` works as expected.
+    #[test]
+    fn u8_slice_len_ensurance() {
+        let slice = [0u8; 0].as_slice();
+        assert_eq!(slice.ensure(0), Some(()));
+        assert_eq!(slice.ensure(1), None);
+
+        let slice = [1u8].as_slice();
+        assert_eq!(slice.ensure(0), Some(()));
+        assert_eq!(slice.ensure(1), Some(()));
+        assert_eq!(slice.ensure(2), None);
+    }
+
+    /// Check that we can read various integers from a slice.
+    #[test]
+    fn pod_reading() {
+        macro_rules! test {
+            ($type:ty) => {{
+                let max = <$type>::MAX.to_ne_bytes();
+                let one = (1 as $type).to_ne_bytes();
+
+                let mut data = Vec::new();
+                let () = data.extend_from_slice(&max);
+                let () = data.extend_from_slice(&one);
+                let () = data.extend_from_slice(&[1, 2, 3, 4, 5, 6, 7, 8]);
+
+                let mut raw = data.as_slice();
+                let uword = raw.read_pod::<$type>().unwrap();
+                assert_eq!(uword, <$type>::MAX);
+
+                let uword = raw.read_pod::<$type>().unwrap();
+                assert_eq!(uword, 1);
+            }};
+        }
+
+        test!(i8);
+        test!(u8);
+        test!(i16);
+        test!(u16);
+        test!(i32);
+        test!(u32);
+        test!(i64);
+        test!(u64);
+        test!(i128);
+        test!(u128);
+    }
+
+    /// Test reading of signed and unsigned 16 and 32 bit values against known
+    /// results.
+    #[test]
+    fn word_reading() {
+        let data = [0x7f, 0x85, 0x36, 0xf9];
+        assert_eq!(data.as_slice().read_u16().unwrap(), 0x857f);
+        assert_eq!(data.as_slice().read_i16().unwrap(), -31361);
+        assert_eq!(data.as_slice().read_u32().unwrap(), 0xf936857f);
+        assert_eq!(data.as_slice().read_i32().unwrap(), -113867393);
```

Comment:
these test assume little-endianness, right? Let's make them endian-agnostic to not come back to this in the future?

Reply:
So these checks are existing code, which I just moved from a completely unrelated file. I wanted to keep it around to make sure that we are not regressing. But it's an easy fix, will adjust.